community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -626,7 +623,17 @@ private void afterRollback()
     private void release()
     {
         locks.releaseAll();
-        pool.release( this );
+        if ( terminated )
+        {
+            // This transaction has been externally marked for termination.
+            // Just dispose of this transaction and don't return it to the pool.
+            dispose();
+        }
+        else
+        {
+            // Return this instance to the pool so that another transaction may use it.
+            pool.release( this );
```

Comment:
yep, ideally it would be nice to pool completely save transaction that does not have anything that was produces by previous master/slave related objects.